### ISSUE TITLE
Redundant Building in Verify Action

### ISSUE THREAD

**Issue by developStorm** [simple-icons/simple-icons-website#76]
### Kind of issue 

- [ ] Feature
- [ ] Bug
- [ ] Design
- [x] Other, namely: CI

### Description
It looks like dependencies are [reinstalled](https://github.com/simple-icons/simple-icons-website/runs/2813103103?check_suite_focus=true#step:4:1) and website is [re-built](https://github.com/simple-icons/simple-icons-website/runs/2813103103?check_suite_focus=true#step:6:8) in "Test Website" action. Therefore, it seems unnecessary to set "Build website" as a precondition for "Test Website", or maybe the whole "Build website" is an unnecessary test (already covered by other ones).

**Comment by ericcornelissen** [comment]
The current design is intentional, though there is indeed some duplication. The "Build Website" job is indeed a precondition for the "Test Website" job as in order to test the website it must be possible to build the website. So, having the build job be a precondition for the test job ensures the test job will only fail because the tests fail and not because of a build error. You might be in favour of putting it in one job, I'm personally of the opinion that it's nice to be able to tell what went wrong without having to go into the logs.

We could definitely look into carrying over the website build between the two jobs, e.g. with caching or artefacts, but with the build taking just 5 seconds I'm not convinced that would be particularly useful. What do you think @developStorm ?

**Comment by developStorm** [comment]
I see what you mean, I also think it may not be worthwhile to add cache for build in the existing situation. 

I am a little curious though, is there a reason you choose to let Publish Website independent of Verify? If I were to design it, I would probably hold the publish until checks pass:
![image](https://user-images.githubusercontent.com/59678453/121956815-6c9af000-cd16-11eb-940a-a09550b841c5.png)

If we go this way, then cache build phase may become actually useful -- But of course, that's a bit weird - like artificially increasing the complexity just to add caching hahaha.

**Comment by developStorm** [comment]
> I'm personally of the opinion that it's nice to be able to tell what went wrong without having to go into the logs.

Having said that, if having clear status is all needed, I feel it is totally ok to run Build and Test in parallel (removing the dependency), so that the total time for waiting Action result to show out can be shortened somewhat

**Comment by ericcornelissen** [comment]
> I see what you mean, I also think it may not be worthwhile to add cache for build in the existing situation.
> 
> I am a little curious though, is there a reason you choose to let Publish Website independent of Verify? If I were to design it, I would probably hold the publish until checks pass:
> ![image](https://user-images.githubusercontent.com/59678453/121956815-6c9af000-cd16-11eb-940a-a09550b841c5.png)
> 
> If we go this way, then cache build phase may become actually useful -- But of course, that's a bit weird - like artificially increasing the complexity just to add caching hahaha.

The "Publish Website" job is in a separate workflow file because it should not always be run. The "Build website", "Test website" and "Lint" jobs should verify all commits and Pull Requests, whereas we only want to publish the website in certain circumstances, currently whenever there's a commit on the `master` branch.

We _could_ merge the workflow files and conditionally run the "Publish Website" job and, as you said, make even more use of the cached build. However, that would mean there is a skipped build in every Pull Request which seems a bit artificial as well.

On top of that, but this is pretty minor, the verify jobs can use [a dev build](https://github.com/simple-icons/simple-icons-website/blob/aea5cb9dfa13c08158110fad5f78894e91ccc58c/package.json#L17) whereas the publishing job should use [a production build](https://github.com/simple-icons/simple-icons-website/blob/master/package.json#L16).


> Having said that, if having clear status is all needed, I feel it is totally ok to run Build and Test in parallel (removing the dependency), so that the total time for waiting Action result to show out can be shortened somewhat

Running them in parallel has, to an extend, the same problem. Of course, if you know what's going on you may realize that if the build job failed than the test job failed for the same reason. But that requires you to understand the build process AND make the connection when you encounter a failed built.

If the concern is the overall time of the build, I simply don't feel the same way - especially considering that this is not a super busy project. For that reason I'm not going to change it based solely on this issue. But, if there's others that feel the same way I'm definitely willing to merge the build and test jobs.

That being said, currently the distribution of time throughout the verify jobs looks approximately like this (based on [this build](https://github.com/simple-icons/simple-icons-website/pull/74/checks?check_run_id=2701720410)):

| # | Duration | Step |
| --- | --- | --- |
| 1 | 31s | Run integration tests |
| 2 | 18s | Install dependencies |
| 3 | 8s | Run unit tests |
| 4 | 5s | Build website |
| 4 | 3s | Run linter |

Based on which I'd say the easiest way to increase the overall CI runtime is by looking into caching dependencies.

**Comment by developStorm** [comment]
Thanks for all these details @ericcornelissen! I've decided to give up on this idea, you made a good point.

BTW, caching dependencies (npm) will insignificantly speed up the workflow for about ~6s: [with cache](https://github.com/developStorm/simple-badges/runs/2823483550?check_suite_focus=true#step:5:9), [w/o cache](https://github.com/simple-icons/simple-icons-website/pull/74/checks?check_run_id=2701720410#step:4:14). I don't think it's very helpful as well, but it's pretty easy to set up, so maybe not bad to try.

**Comment by ericcornelissen** [comment]
And thank you for testing the caching out!

While not much, it's more time than the time it takes to build the website :upside_down_face: I'll look into actually adding caching to the CI builds for this project so I'll leave this open to remind myself of that.

**Comment by developStorm** [comment]
I've already implemented it before so just submit the changes to you 😺. This is done exactly according to the [official instruction](https://github.com/actions/cache/blob/main/examples.md#node---npm).
